File: services/backend/blue-magma-api/observability/observability.go
package observability

import (
	"context"
	"os"
	"strconv"
	"strings"

	"github.com/gofiber/contrib/otelfiber"
	"github.com/gofiber/fiber/v2"
	"github.com/redis/go-redis/extra/redisotel/v9"
	"github.com/redis/go-redis/v9"
	log "github.com/sirupsen/logrus"
	gormotel "go.bryk.io/pkg/otel/gorm"
	"gorm.io/gorm"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
)

// Init configures OpenTelemetry tracing and propagation for the service.
// It is vendor-neutral and only depends on OTLP/HTTP and standard OTEL_* env vars.
func Init(ctx context.Context) (func(context.Context) error, error) {
	serviceName := os.Getenv("OTEL_SERVICE_NAME")
	if serviceName == "" {
		serviceName = "blue-magma-api"
	}

	env := os.Getenv("OTEL_ENVIRONMENT")
	if env == "" {
		env = os.Getenv("ENVIRONMENT")
	}
	if env == "" {
		env = "development"
	}

	res, err := resource.New(
		ctx,
		resource.WithFromEnv(),
		resource.WithTelemetrySDK(),
		resource.WithHost(),
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
			semconv.DeploymentEnvironment(env),
		),
	)
	if err != nil {
		log.WithError(err).Warn("failed to create OpenTelemetry resource, using default")
		res = resource.Default()
	}

	// Default to always sampling if an exporter is configured; otherwise never sample.
	sampler := sdktrace.AlwaysSample()
	if strings.EqualFold(os.Getenv("OTEL_TRACES_SAMPLER"), "parentbased_traceidratio") {
		if arg := os.Getenv("OTEL_TRACES_SAMPLER_ARG"); arg != "" {
			if ratio, err := strconv.ParseFloat(arg, 64); err == nil {
				sampler = sdktrace.ParentBased(sdktrace.TraceIDRatioBased(ratio))
			}
		}
	}

	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")

	var tp *sdktrace.TracerProvider
	var mp *sdkmetric.MeterProvider
	shutdownFuncs := []func(context.Context) error{}

	if endpoint == "" {
		// No endpoint configured: install lightweight providers that never export.
		tp = sdktrace.NewTracerProvider(
			sdktrace.WithSampler(sdktrace.NeverSample()),
			sdktrace.WithResource(res),
		)
		mp = sdkmetric.NewMeterProvider(
			sdkmetric.WithResource(res),
		)
	} else {
		traceClientOpts := []otlptracehttp.Option{
			otlptracehttp.WithEndpoint(endpoint),
		}
		metricClientOpts := []otlpmetrichttp.Option{
			otlpmetrichttp.WithEndpoint(endpoint),
		}

		insecure := os.Getenv("OTEL_EXPORTER_OTLP_INSECURE")
		if insecure == "true" || insecure == "1" {
			traceClientOpts = append(traceClientOpts, otlptracehttp.WithInsecure())
			metricClientOpts = append(metricClientOpts, otlpmetrichttp.WithInsecure())
		}

		traceExporter, err := otlptracehttp.New(ctx, traceClientOpts...)
		if err != nil {
			return nil, err
		}

		metricExporter, err := otlpmetrichttp.New(ctx, metricClientOpts...)
		if err != nil {
			return nil, err
		}

		reader := sdkmetric.NewPeriodicReader(metricExporter)
		mp = sdkmetric.NewMeterProvider(
			sdkmetric.WithReader(reader),
			sdkmetric.WithResource(res),
		)

		tp = sdktrace.NewTracerProvider(
			sdktrace.WithBatcher(traceExporter),
			sdktrace.WithResource(res),
			sdktrace.WithSampler(sampler),
		)

		shutdownFuncs = append(shutdownFuncs, traceExporter.Shutdown, mp.Shutdown)
	}

	otel.SetTracerProvider(tp)
	if mp != nil {
		otel.SetMeterProvider(mp)
	}
	otel.SetTextMapPropagator(
		propagation.NewCompositeTextMapPropagator(
			propagation.TraceContext{},
			propagation.Baggage{},
		),
	)

	// Compose a shutdown that attempts all component shutdowns and returns
	// the first error (if any).
	return func(c context.Context) error {
		// Always attempt to shut down the tracer provider.
		var firstErr error
		if err := tp.Shutdown(c); err != nil && firstErr == nil {
			firstErr = err
		}
		for _, fn := range shutdownFuncs {
			if err := fn(c); err != nil && firstErr == nil {
				firstErr = err
			}
		}
		return firstErr
	}, nil
}

// InstrumentFiber adds OpenTelemetry tracing and metrics to the Fiber app.
func InstrumentFiber(app *fiber.App) {
	if app == nil {
		return
	}
	app.Use(otelfiber.Middleware())
}

// InstrumentGorm attaches OpenTelemetry instrumentation to the given GORM DB.
func InstrumentGorm(db *gorm.DB, dbName string) {
	if db == nil {
		return
	}

	plugin := gormotel.Plugin(gormotel.WithDBName(dbName))
	if err := db.Use(plugin); err != nil {
		log.WithError(err).Warn("failed to register GORM OpenTelemetry plugin")
	}
}

// InstrumentRedisClient enables OpenTelemetry tracing and metrics for the Redis client.
func InstrumentRedisClient(client *redis.Client) {
	if client == nil {
		return
	}

	if err := redisotel.InstrumentTracing(client); err != nil {
		log.WithError(err).Warn("failed to enable Redis tracing")
	}

	if err := redisotel.InstrumentMetrics(client); err != nil {
		log.WithError(err).Warn("failed to enable Redis metrics")
	}
}

File: services/agent/observability.py
import os
from typing import Callable, Optional

from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.propagate import set_global_textmap
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.sampling import TraceIdRatioBased, ParentBased, ALWAYS_ON
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.instrumentation.redis import RedisInstrumentor


def _build_resource() -> Resource:
    service_name = os.getenv("OTEL_SERVICE_NAME", "graphlang-agent")
    env = os.getenv("OTEL_ENVIRONMENT") or os.getenv("ENVIRONMENT") or "development"

    return Resource.create(
        {
            "service.name": service_name,
            "deployment.environment": env,
        }
    )


def _build_sampler():
    sampler_name = os.getenv("OTEL_TRACES_SAMPLER", "").lower()
    if sampler_name == "parentbased_traceidratio":
        arg = os.getenv("OTEL_TRACES_SAMPLER_ARG") or "1.0"
        try:
            ratio = float(arg)
        except ValueError:
            ratio = 1.0
        return ParentBased(TraceIdRatioBased(ratio))
    return ALWAYS_ON


def init_telemetry() -> Callable[[], None]:
    """Initialize OpenTelemetry tracing for the agent.

    This is vendor-neutral: configuration is driven by OTEL_* environment
    variables and data is exported via OTLP/HTTP to the collector.
    """

    resource = _build_resource()
    sampler = _build_sampler()

    endpoint = os.getenv("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT") or os.getenv(
        "OTEL_EXPORTER_OTLP_ENDPOINT"
    )

    # If no endpoint is configured, install a provider without exporters so
    # that instrumentation is active but nothing is sent anywhere.
    if not endpoint:
        provider = TracerProvider(resource=resource, sampler=sampler)
        trace.set_tracer_provider(provider)
        set_global_textmap(TraceContextTextMapPropagator())
        return lambda: None

    if not endpoint.startswith("http"):
        # Assume host:port and build the OTLP/HTTP traces endpoint.
        endpoint = f"http://{endpoint}/v1/traces"

    # For the HTTP OTLP exporter, TLS vs non-TLS is determined by the
    # endpoint scheme (http/https). The 'insecure' flag is only relevant for
    # the gRPC exporter, so we don't pass it here.
    exporter = OTLPSpanExporter(endpoint=endpoint)

    provider = TracerProvider(resource=resource, sampler=sampler)
    processor = BatchSpanProcessor(exporter)
    provider.add_span_processor(processor)

    trace.set_tracer_provider(provider)
    set_global_textmap(TraceContextTextMapPropagator())

    def shutdown() -> None:
        provider.shutdown()

    return shutdown


def instrument_fastapi(app: Optional[FastAPI]) -> None:
    """Attach FastAPI/ASGI instrumentation to the given app."""

    if app is None:
        return
    FastAPIInstrumentor.instrument_app(app)


def instrument_clients() -> None:
    """Instrument common client libraries used by the agent.

    - requests: outbound HTTP calls to the backend API
    - redis: Redis session persistence
    """

    RequestsInstrumentor().instrument()
    RedisInstrumentor().instrument()

